lektor/utils.py

```diff
@@ -529,28 +533,74 @@ def bool_from_string(val, default=None):
     return default
 
 
-def make_relative_url(base, target):
-    """Returns a relative URL from base to target."""
-    if base == '/':
-        depth = 0
-        prefix = './'
+def make_relative_url(source, target):
+    """
+    Returns the relative path (url) needed to navigate
+    from `source` to `target`.
+    """
+
+    # WARNING: this logic makes some unwarranted assumptions about
+    # what is a directory and what isn't. Ideally, this function
+    # would be aware of the actual filesystem.
+    s_is_dir = source.endswith("/")
+    t_is_dir = target.endswith("/")
+
+    source = PurePosixPath(posixpath.normpath(source))
+    target = PurePosixPath(posixpath.normpath(target))
+
+    if not s_is_dir:
+        source = source.parent
+
+    relpath = str(get_relative_path(source, target))
+    if t_is_dir:
+        relpath += "/"
+
+    return relpath
+
+
+def get_relative_path(source, target):
+    """
+    Returns the relative path needed to navigate from `source` to `target`.
+
+    get_relative_path(source: PurePosixPath,
+                      target: PurePosixPath) -> PurePosixPath
+    """
+
+    if not source.is_absolute() and target.is_absolute():
+        raise ValueError("Cannot navigate from a relative path"
+                         " to an absolute one")
+
+    if source.is_absolute() and not target.is_absolute():
+        # nothing to do
+        return target
+
+    if source.is_absolute() and target.is_absolute():
+        # convert them to relative paths to simplify the logic
+        source = source.relative_to("/")
+        target = target.relative_to("/")
+
+    # is the source an ancestor of the target?
+    try:
+        relpath = target.relative_to(source)
+    except ValueError:
+        pass
     else:
-        depth = ('/' + base.strip('/')).count('/')
-        prefix = ''
-        # If the last part of the base path contains a dot, the page will
-        # render into, for example "404.html" instead of "404/index.html".
-        # Any relative links from it therefore need one less level of depth.
-        if '.' in base.strip('/').split('/')[-1]:
-            depth -= 1
-            if depth == 0:
-                prefix = './'
-
-    ends_in_slash = target[-1:] == '/'
-    target = posixpath.normpath(posixpath.join(base, target))
-    if ends_in_slash and target[-1:] != '/':
-        target += '/'
-
-    return (prefix + '../' * depth).rstrip('/') + target
+        return relpath
+
+    # even if it isn't, one of the source's ancestors might be
+    distance = PurePosixPath(".")
+    for ancestor in source.parents:
```

Comment:
Feel free to ignore this if you think it makes things more complicated, but you could do something like:

```python
# Traverse upwards to the closest common ancestor. This will always return since the top is guaranteed to be "."
for distance, ancestor in enumerate(chain([source], source.parents)):
    try:
        return PurePosixPath(*[".."] * distance) / target.relative_to(ancestor)
    except ValueError:
        pass
```

That way you'd avoid having basically the same logic twice, at the expense of some readability.

Reply:
The original was enumerate-based. To me this the version now is highly readable, I think this is pushing it just for the sake of returning from the `try`.

Hmm. How about returning from the `else`?

Reply:
Sure, I think this is OK.